### PR TITLE
Plans (Storage): Separate plan features from add-ons, holistically. Additional refactors

### PR DESCRIPTION
**Pull request by chriskmnds**
Related to #

## Proposed Changes

The main change here revolves around separating the storage associated with a plan as a feature (which equates to the default storage the plan comes with), and the upgrade/add-on that may be available for purchase.

We previously intermingled these concepts at the data layer, defining the `StorageOption` property (now removed) on the plan definitions that included both (storage feature and storage add-ons). 

The other changes involve a bit of cleanup:

- Renaming some files (to also make the distinction clearer e.g. we had a "add-on dropdown" whilst it was a dropdown of both "plan storage feature" and "available add-ons")
- Some CSS class renames to go along with the above
- Deleting redundant code
- Migrates some of the logic to AddOns data-store (`useAvailableStorageAddOns`)

In follow-up PRs, we will separate these concepts even further, and rework the add-ons' logic to provide an indexed structure via add-on slugs (hence stop referencing storage add-on slugs in their `featureSlugs` properties). 

## Why are these changes being made?

Main motivation: 

The storage add-ons are really a platform/UI concern, not a plan property. The only related property that a plan can/should have is (1) the default storage that it comes with and (2) any upper limit on the amount of storage it could expand on. The UI can then serve products for that difference in 50GB, or 100GB, or 99GB, or any other increment or variable amount it desires. In principle/theory, this would be the more flexible approach, while keeping the data layer closer to definition - and further from client features. 

## Testing Instructions

* Ensure storage add-ons work as previously - `/start/plans`
* Ensure current plan in `/plans/[ site ]` is upgradeable with storage once (upgrade , then revisit `/plans` - it should show the upgraded volume along with additional cost)
* Ensure storage is automatically added to the cart when in query param e.g. `/start/business/?storage=50gb-storage-add-on`

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by jeyip** [comment]
Hey @chriskmnds! Pretty busy today but will make sure to take a look at this tomorrow morning 👍

**Comment by chriskmnds** [comment]
> Hey @chriskmnds! Pretty busy today but will make sure to take a look at this tomorrow morning 👍

Thanks @jeyip . 

If I can get a bit of help with testing, then that would be really helpful. It's obviously quite prone to conflicts with 20+ file changes. I'm building on top of this more refactors 😬

**Comment by jeyip** [comment]
Hey Christos! Thanks for the patience and apologies for the delay. I'll be testing this after dinner tonight 🙂

**Comment by jeyip** [comment]
## Testing
### Requirements

For each of the flows listed below:
- [ ] Ensure storage add-ons work as previously - /start/plans
- [ ] Ensure current plan in /plans/[ site ] is upgradeable with storage once (upgrade , then revisit /plans - it should show the upgraded volume along with additional cost)
- [ ] Ensure storage is automatically added to the cart when in query param e.g. /start/business/?storage=50gb-storage-add-on

### Browsers
- [x] Chrome

**Review by jeyip — COMMENTED** [review]
Started code reviewing but running out of steam tonight. Will continue reviewing first thing tomorrow.

**Comment by chriskmnds** [comment]
Thank you @jeyip ! I really appreciate your insightful testing, always 😊

> After upgrading a plan with storage add-ons through /plans and revisiting /plans, the behaviour of the dropdown seems odd. Below is a GIF where I've purchased a 50GB storage add-on upgrade. When I revisit /plans, it defaults to the 150GB add-on option, and when I select the 100GB option, the 150GB add-on option disappears.

Thanks. It was missing a check on "available storage add-ons", so it was rendering the dropdown irrespective of purchased/not-purchased (and getting messed up too). It should be fixed in the last updates. Also extracted a `useAvailableStorageAddons` hook into the `AddOns` data-store now and imported it from there. (will refactor next the add-ons section to also use this same hook)

**Comment by jeyip** [comment]
> Thanks. It was missing a check on "available storage add-ons", so it was rendering the dropdown irrespective of purchased/not-purchased (and getting messed up too). It should be fixed in the last updates. Also extracted a useAvailableStorageAddons hook into the AddOns data-store now and imported it from there. (will refactor next the add-ons section to also use this same hook)

Confirmed that the bug is fixed :D

File: packages/calypso-products/src/plans-list.tsx [hunk 1]
```diff
@@ -673,14 +670,7 @@ const getPlanFreeDetails = (): IncompleteWPcomPlan => ( {
 					FEATURE_SITE_ACTIVITY_LOG_JP,
 					FEATURE_SHARES_SOCIAL_MEDIA_JP,
 			  ],
-	get2023PricingGridSignupStorageOptions: () => {
```
Comment (jeyip) [inline packages/calypso-products/src/plans-list.tsx:676]: `Non-blocking / no response needed`:

Thanks for slowly disentangling the storage add-on from the concept of features 🎉

**Review by jeyip — APPROVED** [review]
There are quite a large volume of changes 😆

I did my best to go through the files and understand the intention behind each of your updates. I didn't see anything concerning. In fact, more than anything, the changes bring lots of clarity to the structure and organization of storage add-ons as they relate to the plans grid.

- Thoughtful renaming of the dropdown 👍 
- Happy to see the removal of tangled storage add-on references and features from within the `plans-list`, the `storage-feature` component, scss file, etc.

Things feel way more straightforward, and I'm excited to see this in production. Besides updating the location of the `use-available-storage-add-ons` hook, not sure if you have anything else planned for this PR, but the code looks great to me.

**Comment by chriskmnds** [comment]
> I was really excited when I saw the logic moved into there 😆 It feels way more intuitive, but I understand why we need to move it back.

Thanks @jeyip ! We can actually keep the `useAvailableStorageAddOns` hook under the add-ons data-store. Just the `ELIGIBLE_PLANS_FOR_STORAGE_UPGRADE` variable is grid-specific right now, so moved that out (which is a candidate for refactoring next). I will try and incorporate the hook in the main add-ons selector at some point. It might end up being renamed or refactored later - but for now, it's a generic function that returns "available storage add-ons for purchase".

(keeping you happy 😃 )

**Pull request merged**